### PR TITLE
Bignum: extract core_mul from the prototype

### LINKED ISSUE #6226: Bignum: extract core_mul from the prototype
The task is to extract `core_mul` from the [prototype](https://github.com/hanno-arm/mbedtls/tree/ecp_prototype):
https://github.com/hanno-arm/mbedtls/blob/ecp_prototype/library/bignum_core.c#L288-L295

There should be separate commits containing the code from the prototype, with the absolute minimum of modifications that make the library compile. (It is Ok if this function is not called at all at this point.) These commits should have Hanno as the author (git commit --author="Hanno Becker <hanno.becker@arm.com>")

The function declaration should go to `library/bignum_core.h` and implementation to `library/bignum_core.c`.

The task includes making the legacy Bignum functions call the extracted functions where the functionality is duplicated. (To minimise impact on code size.)

The prototype is too macro heavy. Most new macros should be expanded/removed or replaced with static functions if possible. (Pre-existing macros should not be touched.) The macro `MPI_CORE` should expand to `mbedtls_mpi_core_ ## func ##` instead of `mbedtls_mpi_core_ ## func ## _minimal`.

### PR DESCRIPTION
**Pull request by tom-cosgrove-arm**
## Description
Extract MPI_CORE(mul) from the [prototype](https://github.com/hanno-arm/mbedtls/blob/ecp_prototype/library/bignum_core.c#L288-L295). Fixes #6226

## Status
**READY**

## Requires Backporting
**NO**

## Todos
- [x] The `MPI_CORE(mul)` and its dependencies are taken from the prototype and credit is given to Hanno
- [x] New macros are expanded/removed
- [x] The legacy function in `bignum.c` (`mbedtls_mpi_mul_mpi()`) is calling the new function wherever possible
- [x] Extensive unit tests are added for all new functions

## Gatekeeper checklist

- [x] **changelog** not required: not user facing
- [x] **backport** not required: new API
- [x] **tests** present: generated unit tests added

### CODE REVIEW

**Comment by tom-cosgrove-arm** [comment]
Unit tests added; this is now ready for review

**Comment by tom-cosgrove-arm** [comment]
Rebased on top of `montmul`, which has had review comments applied. Still the actual work in this PR is only the last 3 commits.

**Comment by tom-cosgrove-arm** [comment]
Rebased on top of the montmul PR https://github.com/Mbed-TLS/mbedtls/pull/6083, which has had a doc comment fix applied. Still the actual work in this PR is only the last 3 commits.

**Review by minosgalanakis — COMMENTED** [review]
Commenting on [last 4 commits](https://github.com/Mbed-TLS/mbedtls/pull/6233/files/b2c06f4acfa55c10fd59a0612a08d68469291041..d4f6663f4dd855e5fd00e99b8097e98255733697).

Looks good

**Comment by tom-cosgrove-arm** [comment]
This is now rebased on top of [the latest montmul PR 6083](https://github.com/Mbed-TLS/mbedtls/pull/6083). The actual work in this PR is still just the last 3 commits.

**Comment by gilles-peskine-arm** [comment]
Since https://github.com/Mbed-TLS/mbedtls/pull/6083/commits is merged, please rebase.

File: tests/suites/test_suite_bignum_core.misc.data [hunk 1]
```diff
@@ -491,6 +491,834 @@ mpi_core_fill_random:42:0:-1:0:MBEDTLS_ERR_MPI_BAD_INPUT_DATA
 Fill random core: 42 bytes, 5 missing limbs
 mpi_core_fill_random:42:0:-5:0:MBEDTLS_ERR_MPI_BAD_INPUT_DATA
 
+mbedtls_mpi_core_mul #1: 0x0 * 0x0 = 0x0
+mpi_core_mul:"0":"0":"0"
```
Comment (tom-cosgrove-arm) [inline tests/suites/test_suite_bignum_core.misc.data:495]: Should these test cases be generated by the Python script now? Or should we raise an issue to do that "later"?

Comment (gabor-mezei-arm) [reply tests/suites/test_suite_bignum_core.misc.data:495]: Yes, it should. It was only a rebase.

**Review by minosgalanakis — COMMENTED** [review]
Code looks ok, except a minor style issue.

It currently tests combinations of the default inputs, including aliasing, which is sufficient.

~~~~~~
        "0", "1", # corner cases
        "2", "3", # small primes
        "4",      # non-prime even
        "38",     # small random
        SAFE_PRIME_192_BIT_SEED_1,  # prime
        RANDOM_192_BIT_SEED_2_NO1,  # not a prime
        RANDOM_192_BIT_SEED_2_NO2,  # not a prime
        SAFE_PRIME_1024_BIT_SEED_3, # prime
        RANDOM_1024_BIT_SEED_4_NO1, # not a prime
        RANDOM_1024_BIT_SEED_4_NO3, # not a prime
        RANDOM_1024_BIT_SEED_4_NO2, # largest (not a prime)
~~~~~~

**Review by minosgalanakis — APPROVED** [review]
LGTM ( Assuming a green CI )

File: library/bignum_core.h [hunk 2]
```diff
@@ -398,6 +398,25 @@ mbedtls_mpi_uint mbedtls_mpi_core_mla(mbedtls_mpi_uint *X, size_t X_limbs,
                                       const mbedtls_mpi_uint *A, size_t A_limbs,
                                       mbedtls_mpi_uint b);
 
+/**
+ * \brief Perform a known-size multiplication
+ *
+ * \p A may be aliased to \p B.
```
Comment (yanesca) [inline library/bignum_core.h:404]: By default we promise that the output may be aliased to one or more inputs: https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.4.0/library/bignum_core.h#L44-L47.

We should explicitly state that X may not be aliased to any of the inputs for this function.

**Review by yanesca — CHANGES_REQUESTED** [review]
Have two minor suggestions for improvement, but other than that it looks good to me.

**Review by yanesca — APPROVED** [review]
LGTM

**Review by minosgalanakis — APPROVED** [review]
LGTM

**Pull request merged**